# Group-level satistical analysis of brain electrical properties  - Diagnostic tests
This notebooks provides the diagnostic tests performed to assess the robustness of the models used in the group-level satistical analysis of brain electrical properties ($\sigma$ and $\epsilon_r$).

Part of nano-eptk package by Arnaud Boutillon (arnaud.boutillon@kcl.ac.uk).

## Python environment setting

In [1]:
import os
import sys
import warnings

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import statsmodels.api as sm

from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.tools.tools import add_constant
from statsmodels.stats.diagnostic import het_breuschpagan, het_white

base_dir = os.path.abspath(os.path.join(os.path.pardir, os.path.pardir))
data_dir = os.path.join(base_dir, "data")
results_dir = os.path.join(base_dir, "results")

warnings.filterwarnings("ignore")

## 1. Neonatal data
Diagnostic tests were performed on subjects whose postmenstrual age (PMA) $\geq$ 37 weeks so that both term-born and preterm infants are considered at similar age at scan; $n = 534$ subjects were considered in total. 

In [2]:
def filter_dHCP_df(input_df_path, output_df_path):
    # Remove NaN and PMA < 37
    df = pd.read_csv(input_df_path)
    df = df.dropna(subset=["brain_MD", "RVV"])
    df = df.drop(df[df["PMA"] < 37].index)
    df.to_csv(output_df_path, index=False)

In [3]:
# Load dataframe
csv_dir = os.path.join(data_dir, "cohorts_csv")
unfiltered_df_path = os.path.join(csv_dir, "dHCP.csv")
df_path = os.path.join(csv_dir, "dHCP_filtered_MD_RVV_PMA.csv")

# Filter and load dataframe
filter_dHCP_df(unfiltered_df_path, df_path)
df = pd.read_csv(df_path) 

# Display cohort size
print(f"dHCP sessions n = {len(df.index)}")
print(f"dHCP subjects m = {df['subject_id'].nunique()}")

dHCP sessions n = 534
dHCP subjects m = 534


## 2. Variance inflation factors (VIFs)
We computed the variance inflation factors (VIF) to measure the magnitude of multicollinearity between regressors: postmenstrual age (PMA), preterm birth (PTB), relative ventricle volume (RVV) and mean diffusivity (MD). All calculated VIF values were below the commonly accepted threshold of $5$, suggesting limited multicollinearity and reliable estimation of regression coefficients.

In [4]:
# Load dataframe
df = pd.read_csv(df_path)

# Data header
x_cols = ["PMA", "RVV", "brain_MD", "PTB"]

# For VIF
df = df[x_cols]
df = add_constant(df)

# Iter over regressors
for i in range(1, df.shape[1]):

    # Compute and display VIF
    vif = variance_inflation_factor(df.values, i)
    print(f"VIF {df.columns[i]}: {vif:.2f}")

VIF PMA: 1.37
VIF RVV: 1.24
VIF brain_MD: 1.66
VIF PTB: 1.51


## 3. Non-linear and interaction terms
Interaction terms between predictors were included in the multivariate model to assess, for example, whether prematurity had an non-linear effect on the relationship between conductivity and development (i.e. $\textrm{PMA} \times \textrm{PTB}$ interaction). Similarly, we tried incorporating quadratic and cubic terms (e.g. $\textrm{PMA}^2$ or $\textrm{PMA}^3$) in our models. However, none of these non-linear interaction resulted in a better model fit, so these interaction terms were discarded from the statistical analysis.

In [5]:
# Set file path
summary_dir = os.path.join(results_dir, "statsmodels")
summary_path = os.path.join(summary_dir, "diagnostic_experiment1.txt")

# Load dataframe
df = pd.read_csv(df_path) 

# Data header
y_col = "brain_SIG"
x_cols = ["PMA", "RVV", "brain_MD", "PTB"]

# Center data
for x_col in x_cols:
    df[x_col] -= df[x_col].mean()

# Fit linear model
formula = f"{y_col} ~  {' + '.join(x_cols)}"
lm = smf.ols(formula=formula, data=df).fit()

# Fit linear model with interaction
formula_inter = f"{y_col} ~  {' + '.join(x_cols)} + PMA*PTB"
lm_inter = smf.ols(formula=formula_inter, data=df).fit()

# Fit linear model with non-linear term
formula_nl = f"{y_col} ~  {' + '.join(x_cols)} + I(PMA**2)"
lm_nl = smf.ols(formula=formula_nl, data=df).fit()

# Initialize output directory
output_dir = os.path.dirname(summary_path)
os.makedirs(output_dir, exist_ok=True)

# Write summary
with open(summary_path, "w") as fh:
    fh.write(lm.summary(slim=True).as_text())
    fh.write("\n\n")
    fh.write(lm_inter.summary(slim=True).as_text())
    fh.write("\n\n")
    fh.write(lm_nl.summary(slim=True).as_text())

# Summary
print(f"PMAxPTB interaction term significance level: {lm_inter.pvalues["PMA:PTB"]:.3f}")
print(f"PMAxPMA non-linear term significance level:  {lm_nl.pvalues["I(PMA ** 2)"]:.3f}")

PMAxPTB interaction term significance level: 0.230
PMAxPMA non-linear term significance level:  0.458


Both the $\textrm{PMA} \times \textrm{PTB}$ interaction and non-linear $\textrm{PMA}^2$ terms are non-significant (*p* = 0.230 and *p* = 0.458).

## 4. Heteroscedasticity test
Heteroscedasticity tests revealed no evidence of variance heterogeneity in the residuals. Consequently, we did not apply heteroscedasticity correction, as it is rarely used in conjunction with linear mixed-effects (LME) models.

In [6]:
# Load dataframe
df = pd.read_csv(df_path) 

# Data header
y_col = "brain_SIG"
x_cols = ["PMA", "RVV", "brain_MD", "PTB"]

# Center data
for x_col in x_cols:
    df[x_col] -= df[x_col].mean()

# Fit linear model
formula = f"{y_col} ~  {' + '.join(x_cols)}"
lm = smf.ols(formula=formula, data=df).fit()

# Residuals and design matrix (exog)
resid = lm.resid
exog = lm.model.exog

# Breusch-Pagan test
bp_tests = het_breuschpagan(resid, exog)
bp_labels = ['LM stat', 'LM p-value', 'F-stat', 'F p-value']
print("Breusch-Pagan test:")
for bp_label, bp_test in zip(bp_labels, bp_tests):
    print(f"{bp_label}: {bp_test:.4f}")

# White test 
hw_tests = het_white(resid, exog)
hw_labels = ['LM stat', 'LM p-value', 'F-stat', 'F p-value']
print("\nWhite test:")
for hw_label, hw_test in zip(hw_labels, hw_tests):
    print(f"{hw_label}: {hw_test:.4f}")

Breusch-Pagan test:
LM stat: 2.6770
LM p-value: 0.6133
F-stat: 0.6663
F p-value: 0.6156

White test:
LM stat: 20.8324
LM p-value: 0.1060
F-stat: 1.5049
F p-value: 0.1045


## 5. Impact of outliers

### 5.a. Robust linear models
We investigated the impact of outlier data points by employing separate robust linear models, which produced results consistent with those from LMEs. This suggested that any potential outliers were not influential enough to alter the model parameters. In other words, rather than incorporating robust estimators directly into mixed-effects models, we performed a diagnostic check by comparing our LME results to those from robust models.

In [7]:
# Set file path
summary_path = os.path.join(summary_dir, "diagnostic_experiment2.txt")

# Load dataframe
df = pd.read_csv(df_path) 

# Data header
y_col = "brain_SIG"
x_cols = ["PMA", "RVV", "brain_MD", "PTB"]

# Center data
for x_col in x_cols:
    df[x_col] -= df[x_col].mean()

# Fit linear model
formula = f"{y_col} ~  {' + '.join(x_cols)}"
lm = smf.ols(formula=formula, data=df).fit()

# Endog and exog variables
exog = df[x_cols]
exog = add_constant(exog)
endog = df[y_col]

# Fit robust model
rlm = sm.RLM(endog, exog, M=sm.robust.norms.HuberT()).fit()

# Initialize output directory
output_dir = os.path.dirname(summary_path)
os.makedirs(output_dir, exist_ok=True)

# Write summary
with open(summary_path, "w") as fh:
    fh.write(lm.summary(slim=True).as_text())
    fh.write("\n\n")
    fh.write(rlm.summary2().as_text())

# Summary
overlap = lambda a, b: max(0, min(a[1], b[1]) - max(a[0], b[0]))
for i, x_col in enumerate(x_cols):
    CIa = lm.conf_int().loc[x_col]
    CIb = rlm.conf_int().loc[x_col]
    if overlap(CIa, CIb) > 0:
        print(f"{x_col}: 95% CI overlap")
    else:
        print(f"{x_col}: no 95% CI overlap")

PMA: 95% CI overlap
RVV: 95% CI overlap
brain_MD: 95% CI overlap
PTB: 95% CI overlap


### 5.b. Leverage points
To assess the effect of potential leverage points ($x$-extrema points), we excluded points outside the $x$-range $1\%-99\%$ percentiles. The results were consistent with those of the LME models fitted to all data points, suggesting that no leverage point had corrupted our regression analysis.

In [8]:
# Data path
summary_path = os.path.join(summary_dir, "diagnostic_experiment3.txt")

# Load dataframe
df = pd.read_csv(df_path) 

# Date header
y_col = "brain_SIG"
x_cols = ["PMA", "RVV", "brain_MD", "PTB"]

# Center data
for x_col in x_cols:
    df[x_col] -= df[x_col].mean()

# Fit linear model
formula = f"{y_col} ~  {' + '.join(x_cols)}"
lm = smf.ols(formula=formula, data=df).fit()

# 1%-99% percentile for all regressors
q_lows = []
q_his = []
for x_col in x_cols:
    q_lows += [df[x_col].quantile(0.01)]
    q_his += [df[x_col].quantile(0.99)]

# Merge mask
mask = np.ones(len(df), dtype=bool)
for i, x_col in enumerate(x_cols):
    mask &= (df[x_col] > q_lows[i]) & (df[x_col] < q_his[i])
df = df[mask]

# Fit linear model
lm_lp = smf.ols(formula=formula, data=df).fit()

# Initialize output directory
output_dir = os.path.dirname(summary_path)
os.makedirs(output_dir, exist_ok=True)

# Write summary
with open(summary_path, "w") as fh:
    fh.write(lm.summary(slim=True).as_text())
    fh.write("\n\n")
    fh.write(lm_lp.summary(slim=True).as_text())

# Summary
overlap = lambda a, b: max(0, min(a[1], b[1]) - max(a[0], b[0]))
for i, x_col in enumerate(x_cols):
    CIa = lm.conf_int().loc[x_col]
    CIb = lm_lp.conf_int().loc[x_col]
    if overlap(CIa, CIb) > 0:
        print(f"{x_col}: 95% CI overlap")
    else:
        print(f"{x_col}: no 95% CI overlap")

PMA: 95% CI overlap
RVV: 95% CI overlap
brain_MD: 95% CI overlap
PTB: 95% CI overlap


## 6. Radiological scores
Neonatal dHCP structural scans were scored by a radiologist into the following five categories: 
- Normal appearance for age.
- Incidental findings with unlikely significance for clinical outcome or analysis (e.g. subdural haemorrhage. Isolated subependymal cysts. Mild Inferior vermis rotation).
- Incidental findings with possible clinical significance. Unlikely analysis significance (e.g. Isolated non brain anomaly e.g. in pituitary / on tongue).
- Incidental findings with unlikely clinical significance but possible analysis significance (e.g. Several punctate lesions or other focal white matter/ cortical lesions not thought to be of clinical significance).
- Incidental findings with possible/likely significance for both clinical and imaging analysis (e.g. Major lesions within white / matter, cortex, cerebellum and or basal ganglia, small head/brain < 1st centile).

To assess the effect of subjects with incidental findings on regression coefficients, these subjects were excluded before fitting a linear model. As in previous experiments, the results were consistent with those of the LME models fitted to all data points, suggesting that no incidental findings had corrupted our regression analysis.

In [9]:
# Data path
summary_path = os.path.join(summary_dir, "diagnostic_experiment4.txt")

# Load dataframe
df = pd.read_csv(df_path) 

# Date header
y_col = "brain_SIG"
x_cols = ["PMA", "RVV", "brain_MD", "PTB"]

# Center data
for x_col in x_cols:
    df[x_col] -= df[x_col].mean()

# Fit linear model
formula = f"{y_col} ~  {' + '.join(x_cols)}"
lm = smf.ols(formula=formula, data=df).fit()

# 
df = df[df["rad_score"] > 1]

# Fit linear model
lm_rs = smf.ols(formula=formula, data=df).fit()

# Initialize output directory
output_dir = os.path.dirname(summary_path)
os.makedirs(output_dir, exist_ok=True)

# Write summary
with open(summary_path, "w") as fh:
    fh.write(lm.summary(slim=True).as_text())
    fh.write("\n\n")
    fh.write(lm_rs.summary(slim=True).as_text())

# Summary
overlap = lambda a, b: max(0, min(a[1], b[1]) - max(a[0], b[0]))
for i, x_col in enumerate(x_cols):
    CIa = lm.conf_int().loc[x_col]
    CIb = lm_rs.conf_int().loc[x_col]
    if overlap(CIa, CIb) > 0:
        print(f"{x_col}: 95% CI overlap")
    else:
        print(f"{x_col}: no 95% CI overlap")

PMA: 95% CI overlap
RVV: 95% CI overlap
brain_MD: 95% CI overlap
PTB: 95% CI overlap
